File: dotnet/NjnuClassroom/SearchMore.cs
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Primitives;
using ZTxLib.NETCore.DaoTemplate;
using ZTxLib.NETCore.DaoTemplate.MySQL;

namespace NjnuClassroom
{
    internal static class SearchMore
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="context"></param>
        /// <param name="parameters"></param>
        /// <returns></returns>
        public static async Task ProcessRequest(HttpContext context,
            ImmutableDictionary<string, StringValues> parameters)
        {
            if (SettingService.ConfigurationSettings["service"] == "off")
            {
                await context.Response.WriteAsync(
                    "{" +
                    "\"status\":1," +
                    "\"message\":\"service off\"," +
                    $"\"service\":\"off\"," +
                    $"\"data\":[]" +
                    "}");
                return;
            }

            var dao = SettingService.Dao;
            string[] days = {"sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"};

            if (new[] {"day", "jc_ks", "jc_js", "jxl", "zylxdm", "kcm"}.Any(key => !parameters.ContainsKey(key)))
            {
                context.Response.StatusCode = 400;
                return;
            }

            var day = parameters["day"];
            var jc_ks = parameters["jc_ks"];
            var jc_js = parameters["jc_js"];
            var jxl = parameters["jxl"];
            var zylxdm = parameters["zylxdm"];
            var kcm = parameters["kcm"];
            if (kcm == "#") kcm = "_";
            if (day != "#") days = new[] {days[int.Parse(day)]};
            var jc = $"(`jc_ks`>={jc_ks} AND `jc_js`<={jc_js})";
            jxl = jxl == "#" ? "(`jxl` IS NOT NULL)" : $"(`jxl`='{jxl}')";
            zylxdm = zylxdm == "#" ? "(`zylxdm` IS NOT NULL)" : $"(`zylxdm`='{zylxdm}')";

            var classrooms = new List<Classroom>();
            for (var d = 0; d < days.Length; d++)
            {
                dao.Prepare(
                    "SELECT * FROM `${day}` WHERE ${jc} AND ${jxl} AND ${zylxdm} AND (`jyytms` LIKE #{kcm} OR `kcm` LIKE #{kcm})",
                    concat: new[]
                    {
                        new KvPair("day", days[d]),
                        new KvPair("jc", jc),
                        new KvPair("jxl", jxl),
                        new KvPair("zylxdm", zylxdm),
                    },
                    parameter: new[]
                    {
                        new KvPair("kcm", $"%{kcm}%"),
                    }
                );
                foreach (var classroom in new DataList<Classroom>(new Classroom.RowMapper(), dao))
                {
                    classroom.Day = (day == "#") ? d : int.Parse(day);
                    classrooms.Add(classroom);
                }
            }

            await context.Response.WriteAsync(Format(classrooms));
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="classrooms"></param>
        /// <returns></returns>
        private static string Format(IEnumerable<Classroom> classrooms)
        {
            var id = 0;
            var data =
                $"[{string.Join(",", classrooms.Select(classroom => $"{{{string.Join(",", new Dictionary<string, object> {{"id", ++id}, {"jxl", classroom.Jxl}, {"classroom", classroom.Jsmph}, {"day", classroom.Day}, {"jc_ks", classroom.Jc_ks}, {"jc_js", classroom.Jc_js}, {"zylxdm", classroom.Zylxdm}, {"jyytms", classroom.Jyytms.Replace('\r', '\n').Replace("\n", "")}, {"kcm", classroom.Kcm.Replace('\r', '\n').Replace("\n", "")}}.Select(obj => $"\"{obj.Key}\":\"{obj.Value}\""))}}}"))}]";
            return "{" +
                   "\"status\":0," +
                   "\"message\":\"ok\"," +
                   $"\"service\":\"on\"," +
                   $"\"data\":{data}" +
                   "}";
        }
    }
}

File: flask/handler/SearchMore.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
# @Time     :  2020/10/18 0018
# @Author   :  Zhou Tianxing
# @Software :  PyCharm Professional x64
# @FileName :  SearchMore.py
""""""
import app
import utils

days = ["sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"]
day_mapper = {"sunday": 0, "monday": 1, "tuesday": 2, "wednesday": 3, "thursday": 4, "friday": 5, "saturday": 6}


def handler(args: dict) -> dict:
    if app.config['service'] == 'off':
        return {
            'status': 1,
            'message': "service off",
            'service': "off",
            'data': []
        }

    if 'day' not in args.keys() or args['day'] != "#" and not args['day'].isdigit():
        raise KeyError('day')
    elif 'jc_ks' not in args.keys() or not args['jc_ks'].isdigit():
        raise KeyError('jc_ks')
    elif 'jc_js' not in args.keys() or not args['jc_js'].isdigit():
        raise KeyError('jc_js')
    elif 'jxl' not in args.keys():
        raise KeyError('jxl')
    elif 'zylxdm' not in args.keys():
        raise KeyError('zylxdm')
    elif 'kcm' not in args.keys():
        raise KeyError('kcm')

    jc = "`jc_ks`>=%(jc_ks)s AND `jc_js`<=%(jc_js)s"
    day = True if args['day'] == "#" else f"`day`={days[int(args['day'])]}"
    jxl = True if args['jxl'] == "#" else "`jxl`=%(jxl)s"
    zylxdm = True if args['zylxdm'] == "#" else "`zylxdm`=%(zylxdm)s"
    keyword = "`jyytms` LIKE %(keyword)s OR `kcm` LIKE %(keyword)s"
    args['keyword'] = f"%{args['kcm']}%"

    result = [
        {
            'jasdm': item['JASDM'],
            'JASDM': item['JASDM'],

            'jxl': item['JXLMC'],
            'JXLMC': item['JXLMC'],

            'jsmph': item['jsmph'],

            'capacity': item['SKZWS'],
            'SKZWS': item['SKZWS'],

            'day': day_mapper[item['day']],
            'jc_ks': item['jc_ks'],
            'jc_js': item['jc_js'],

            'zylxdm': item['zylxdm'],
            'jyytms': item['jyytms'],
            'kcm': item['kcm'],
        } for item in
        utils.database.fetchall(
            sql=f"SELECT * FROM `pro` WHERE ({day}) AND ({jc}) AND ({jxl}) AND ({zylxdm}) AND ({keyword})",
            args=args
        )
    ]

    return {
        'status': 0,
        'message': "ok",
        'service': "on",
        'data': result
    }
